cli/azd/cmd/templates.go

```diff
@@ -175,3 +208,290 @@ func getCmdTemplateHelpDescription(*cobra.Command) string {
+	sourceConfigs, err := a.sourceManager.List(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list template sources: %w", err)
+	}
+
+	if a.formatter.Kind() == output.TableFormat {
+		columns := []output.Column{
+			{
+				Heading:       "Key",
+				ValueTemplate: "{{.Key}}",
+			},
+			{
+				Heading:       "Name",
+				ValueTemplate: "{{.Name}}",
+			},
+			{
+				Heading:       "Type",
+				ValueTemplate: "{{.Type}}",
+			},
+			{
+				Heading:       "Location",
+				ValueTemplate: "{{.Location}}",
+			},
+		}
+
+		err = a.formatter.Format(sourceConfigs, a.writer, output.TableFormatterOptions{
+			Columns: columns,
+		})
+	} else {
+		err = a.formatter.Format(sourceConfigs, a.writer, nil)
+	}
+
+	return nil, err
+}
+
+func newTemplateSourceAddCmd() *cobra.Command {
+	return &cobra.Command{
+		Use:   "add <key>",
+		Short: fmt.Sprintf("Adds an azd template source at the specified key %s", output.WithWarningFormat("(Beta)")),
+		Args:  cobra.ExactArgs(1),
+	}
+}
+
+type templateSourceAddFlags struct {
+	name     string
+	location string
+	kind     string
+}
+
+func newTemplateSourceAddFlags(cmd *cobra.Command) *templateSourceAddFlags {
+	flags := &templateSourceAddFlags{}
+
+	cmd.Flags().StringVarP(&flags.kind, "type", "t", "", "Kind of the template source")
+	cmd.Flags().StringVarP(&flags.location, "location", "l", "", "Location of the template source")
+	cmd.Flags().StringVarP(&flags.name, "name", "n", "", "Name of the template source")
+
+	return flags
+}
+
+type templateSourceAddAction struct {
+	flags         *templateSourceAddFlags
+	console       input.Console
+	sourceManager templates.SourceManager
+	args          []string
+}
+
+func newTemplateSourceAddAction(
+	flags *templateSourceAddFlags,
+	console input.Console,
+	sourceManager templates.SourceManager,
+	args []string,
+) actions.Action {
+	return &templateSourceAddAction{
+		flags:         flags,
+		console:       console,
+		sourceManager: sourceManager,
+		args:          args,
+	}
+}
+
+func (a *templateSourceAddAction) Run(ctx context.Context) (*actions.ActionResult, error) {
+	a.console.MessageUxItem(ctx, &ux.MessageTitle{
+		Title: "Add azd template source (azd template source add)",
+	})
+
+	var key = a.args[0]
+	sourceConfig := &templates.SourceConfig{}
+
+	if _, ok := templates.WellKnownSources[key]; !ok {
+		sourceConfig = &templates.SourceConfig{
+			Key:      key,
+			Type:     templates.SourceKind(a.flags.kind),
+			Location: a.flags.location,
+			Name:     a.flags.name,
+		}
+
+		// Validate the custom source config
+		spinnerMessage := "Validating template source"
+		a.console.ShowSpinner(ctx, spinnerMessage, input.Step)
+		_, err := a.sourceManager.CreateSource(ctx, sourceConfig)
+		if err != nil {
+			a.console.StopSpinner(ctx, spinnerMessage, input.StepFailed)
+			return nil, fmt.Errorf("template source validation failed: %w", err)
+		}
+
+		a.console.StopSpinner(ctx, spinnerMessage, input.StepDone)
```

Comment:
consider using `input.GetStepResultFormat()` for StopSpinner

```suggestion
		_, err := a.sourceManager.CreateSource(ctx, sourceConfig)
		a.console.StopSpinner(ctx, spinnerMessage, input.GetStepResultFormat(err))
		if err != nil {
			return nil, fmt.Errorf("template source validation failed: %w", err)
		}
```